File: validation/runtime_memory/validateRuntimeMemory.ts
import {
  buildRuntimeGraph,
  buildRuntimeMemory,
  mergeRuntimeMemories,
  queryRuntimeMemory,
  stableMemoryHash,
} from "../../src/index.js";

const graph = buildRuntimeGraph({ session: { ok: true } });
const mem = buildRuntimeMemory(graph, [{ step: 1, kind: "workflow" }]);
const results = {
  memory_match:
    (mem.stable_hash as string) === stableMemoryHash(graph, [{ step: 1, kind: "workflow" }]),
  query_match: queryRuntimeMemory(mem, "graph") != null,
  merge_match: mergeRuntimeMemories(mem, mem).bounded === true,
  deterministic:
    (mem.stable_hash as string) ===
    (buildRuntimeMemory(graph, [{ step: 1, kind: "workflow" }]).stable_hash as string),
};
const allOk = Object.values(results).every(Boolean);
console.log(allOk ? "PASS" : "FAIL", results);
if (!allOk) process.exit(1);

File: validation/runtime_memory/validate_runtime_memory.dart
import 'dart:convert';
import 'dart:io';

import 'package:webweavex/webweavex.dart';

void main() {
  final graph = buildRuntimeGraph({
    'session': {'ok': true}
  });
  final mem = buildRuntimeMemoryFabric(graph, [
    {'step': 1, 'kind': 'workflow'},
  ]);
  final queried = queryRuntimeMemoryFabric(mem, "graph");
  final merged = mergeRuntimeMemories(mem, mem);

  final results = {
    'memory_match': mem['stable_hash'] ==
        stableMemoryFabricHash(graph, [
          {'step': 1, 'kind': 'workflow'},
        ]),
    'query_match': queried != null,
    'merge_match': merged['stable_hash'] != null && merged['bounded'] == true,
    'deterministic': mem['stable_hash'] ==
        buildRuntimeMemoryFabric(graph, [
          {'step': 1, 'kind': 'workflow'},
        ])['stable_hash'],
  };

  final allOk = results.values.every((x) => x == true);
  Directory('validation/runtime_memory').createSync(recursive: true);
  File('validation/runtime_memory/runtime_memory_vectors.json')
      .writeAsStringSync(
    const JsonEncoder.withIndent('  ').convert({
      'algorithm': 'webweavex-runtime-memory-v2.0.0',
      'vectors': [results],
    }),
  );
  stdout.writeln(allOk ? 'PASS' : 'FAIL');
  if (!allOk) exit(1);
}
